File: trunk/Demo/Demo/Walk.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using System.Windows.Forms;
using RobobuilderLib;

namespace Demo
{

    struct compare
    {
        public int av;
        public int[] dp;

        public compare(int a, int[] c)
        { av = a; dp = c; }
    };

    class Compare
    {
        List<compare> items = new List<compare>();

        public Compare(string f)
        {
            matrix m2 = new matrix(f);

            for (int i = 0; i < m2.getr(); i++)
            {
                double[] t = m2.getrow(i);
                items.Add(new compare((int)vectors.head(t), vectors.convInt(vectors.tail(t))));
            }
        }

        public void close()
        {
            items.Clear();
        }

        public void build()
        {

        }

        public int[] match(int g)
        {
            int[] res = null;
            int diff = 99;
            foreach (compare r in items)
            {
                if (Math.Abs(r.av - g) < diff)
                {
                    diff = Math.Abs(r.av - g);
                    res = r.dp;
                }
            }
            return res;
        }
    }

    class BalanceWalk
    {
        public bool wlk; 
        public int dely = 50;
        public string state = "s"; // "R" // "L" or "r" and "l"

        wckMotion w;
        int n_of_s;
        long st;

        bool dhf = false;

        int gx = 0, gy = 0, gz = 0;

        byte[][] fstep;

        Compare zm;

        public BalanceWalk(wckMotion w1, bool dh)
        {
            w = w1;
            dhf = dh;

            n_of_s = countServos(22);
            Console.WriteLine("Balance walk - {0}", n_of_s);

            matrix m;
            
            if (dhf)
                m = new matrix("rls2.csv");
            else
                m = new matrix("rlstep.csv");

            int r = m.getr();

            fstep = new byte[r][];


            for (int i = 0; i < r; i++)
            {
                if (dhf)
                    fstep[i] = vectors.convByte(m.getrow(i));
                else
                    fstep[i] = cv18(vectors.convByte(m.getrow(i)));
            }

            zm = new Compare("compare.csv");
        }

        byte[] cv18(byte[] a) // hip conversion
        {
            byte[] r = new byte[a.Length];

            for (int i = 0; i < a.Length; i++)
                r[i] = a[i];

            if (n_of_s > 16)
            {
                r[0] += 18;
                r[5] -= 20;
            }
            return r;
        }

        int countServos(int m)
        {
            for (int i = 0; i < m; i++)
            {
                if (!w.wckReadPos(i))
                {
                    return i;
                }
            }
            return m;
        }

        byte[][] reverse(byte[][] z)
        {
            byte[][] r = new byte[z.Length][];

            for (int i = 0; i < z.Length; i++)
                r[i] = z[z.Length - i - 1];

            return r;
        }

        void mirror(bool f)
        {
            byte p1=0,p2=0,p3=0,p4=0,p5=0,p6=0;
            int n = f ? 13 : 10; // reading side
            int m = f ? 10 : 13; // writing side

            w.wckPassive(n);                    
            w.wckPassive(n+1);
            w.wckPassive(n+2);

            if (w.wckReadPos(n))   p1=w.respnse[1];
            if (w.wckReadPos(n+1)) p2=w.respnse[1];
            if (w.wckReadPos(n+2)) p3=w.respnse[1];

            if (w.wckReadPos(m))     p4 = w.respnse[1];
            if (w.wckReadPos(m + 1)) p5 = w.respnse[1];
            if (w.wckReadPos(m + 2)) p6 = w.respnse[1];

            if (p4 != (254 - p1)) w.wckMovePos(m,     254 - p1, 4);
            if (p5 != (254 - p2)) w.wckMovePos(m + 1, 254 - p2, 4);
            if (p6 != (254 - p3)) w.wckMovePos(m + 2, 254 - p3, 4);
        }

        void calibrateXYZ()
        {
            gx = 0; gy = 0; gz = 0;

            if (w.wckReadAll())
            {
                gx = w.cbyte(w.respnse[1]); 
                gy = w.cbyte(w.respnse[2]);
                gz = w.cbyte(w.respnse[3]);
            }
            Console.WriteLine("calibrated: {0},{1},{2}", gx, gy, gz);
        }

        public void motion (Utility u)
        {
            byte[][] cpos=null;
            byte[] sbase=new byte[16] {0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
            int[] az=null;
            int counter=0;

            bool bal = true;

            int x=0, z=0, dz=0;
            int d = 0; 

            calibrateXYZ();

            int nc = 0;
            Double rt = 0; 
            st = DateTime.Now.Ticks;

            wlk = true;

            while (wlk)
            {
                System.Windows.Forms.Application.DoEvents();

                if (w.wckReadAll())
                {
                    x = w.cbyte(w.respnse[1])-gx;
                    int tz = w.cbyte(w.respnse[3])-gz;
                    dz = z - tz;
                    z = tz;
                    d = w.respnse[0];

                    if (w.respnse[4] < 255)
                    {
                        switch((PCremote.RemoCon)(w.respnse[4]))
                        {
                            case PCremote.RemoCon.Stop:
                                if (state == "R")
                                    state = "s";
                                else
                                    wlk = false;
                                break;

                            case PCremote.RemoCon.Forward:
                                if (state == "s")
                                    state = "R";
                                break;

                            case PCremote.RemoCon.Back:
                                if (state == "s")
                                    state = "r";
                                break;

                            case PCremote.RemoCon.A:
                                bal = !bal;
                                break;
                        }
                    }

                    Console.WriteLine("{0},{1},{2},{3},{4},{5}", counter, z, dz, d, w.respnse[4], bal);

                }

                if ((nc++ % 10) == 0)
                {
                    //timing loop
                    rt = (DateTime.Now.Ticks - st) / (10 * TimeSpan.TicksPerMillisecond);
                    st = DateTime.Now.Ticks;
                    w.wckReadPos(30, 5); // for PSD read
                }

                u.pwin(z, dz, nc, rt);

                if (counter == 0)
                {
                    switch (state)
                    {
                        case "R":
                            cpos = fstep;
                            if (d < 15) state = "r";
                            break;

                        case "r":
                            cpos = reverse(fstep);
                            state = "r";
                            if (d > 25) state = "R";
                            break;
                        
                        case "s":
                            cpos = new byte[1][];
                            if (dhf)
                                cpos[0] = wckMotion.basicdh;
                            else
                                cpos[0] = wckMotion.basic18;
                            break;

                        case "m":
                            mirror(true);
                            cpos = null;
                            break;
                    }
                }

                try
                {
                    int[] dp = zm.match(z);

                    if (az == null)
                        az= dp;
                    else
                    {
                        if (dp!=null)
                        {
                            az = vectors.add(az, dp);
                        }

                        Console.Write("Pwr={0}", vectors.rms(az));
                    }

                    if (cpos != null)
                    {
                        if (bal)
                            sbase = vectors.bcheck(vectors.add(az, cpos[counter]), wckMotion.lb_Huno, wckMotion.ub_Huno); // lb_Huno, ub_Huno);
                        else
                            sbase = cpos[counter];

                        w.SyncPosSend(15, 4, sbase, 0);
                        System.Threading.Thread.Sleep(dely);

                        if (++counter >= cpos.Length) counter = 0;
                    }

                    Console.Write(state);
                }
                catch
                {
                    Console.WriteLine("fail");
                    wlk = false;
                }
            }
        }

    }
}

File: Mobile/Android/RoboAndroid/src/net/robobuilderlib/Walk.java
package net.robobuilderlib;

public class Walk
{
    wckMotion w;
    
    boolean first = true;
    boolean hip, dh, done;
    int fdel;
    int fpsd;
    
    Thread	wThread;  

    public Walk(wckMotion a, boolean h, boolean d)
    {
        w = a;
        wThread = null;
        hip=h; dh=d;
    }

    int[][] step = new int[][] {
        new int[] {123, 156, 212,  80, 108, 126, 73, 40, 150, 141,  68, 44, 40, 138, 208, 195},
        new int[] {130, 165, 201,  81, 115, 134, 81, 31, 147, 149,  72, 44, 40, 145, 209, 201},
        new int[] {132, 171, 197,  83, 117, 137, 86, 28, 148, 152,  78, 43, 41, 154, 209, 206},
        new int[] {132, 175, 195,  87, 117, 139, 91, 27, 152, 154,  87, 43, 43, 164, 209, 211},
        new int[] {132, 178, 197,  91, 117, 137, 95, 28, 157, 152,  97, 43, 48, 172, 209, 213},
        new int[] {130, 179, 201,  95, 115, 134, 96, 31, 161, 149, 105, 43, 53, 179, 210, 214},
        new int[] {127, 178, 206,  98, 112, 130, 95, 35, 166, 145, 111, 42, 57, 182, 210, 214},
        new int[] {124, 175, 212, 100, 109, 127, 92, 40, 170, 142, 113, 42, 59, 183, 210, 214},
        new int[] {120, 172, 217, 102, 105, 123, 88, 46, 170, 138, 111, 42, 57, 182, 210, 214},
        new int[] {116, 167, 221, 103, 101, 120, 83, 51, 169, 135, 106, 43, 53, 179, 210, 214},
        new int[] {113, 162, 224, 102,  98, 118, 77, 55, 167, 133, 97,  43, 48, 173, 209, 213},
        new int[] {111, 157, 225, 98,   96, 118, 73, 57, 163, 133, 87,  43, 43, 164, 209, 211},
        new int[] {113, 153, 224, 93,   98, 118, 70, 55, 159, 133, 79,  43, 41, 154, 209, 206},
        new int[] {116, 152, 221, 89,  101, 120, 69, 51, 155, 135, 72,  44, 40, 146, 209, 201},
        new int[] {120, 153, 217, 84,  105, 123, 70, 46, 152, 138, 69,  44, 40, 140, 208, 197},
        new int[] {123, 156, 212, 80,  108, 126, 73, 40, 150, 141, 68,  44, 40, 138, 208, 195}
        };

    int[][] reverse(int[][] z)
    {
        int[][] r = new int[z.length][];

        for (int i = 0; i < z.length; i++)
            r[i] = z[z.length-i-1];

        return r;
    }

    byte[] cv18(int[] a) // hip conversion
    {
        byte[] r = new byte[a.length];       

        for (int i = 0; i < a.length; i++)
        {
        	int v = a[i];
        	
            if (hip && (i==0)) v += 18;
            if (hip && (i==5)) v -= 18;
            
            r[i] = (byte)v;
        }
        
        if (dh)
        {
        	r[15] = r[12]=(byte)255; //
        	r[14] = r[11]=(byte)255; //
        	//r[13] = r[10]=(byte)255; //
        }
        return r;
    }

    void setallLeds(int n, boolean a, boolean b)
    {
        for (int i = 0; i <= n; i++)
        {
            w.wckWriteIO(i, a, b);
        }
    }
    
    public void stop()
    {
        Debug.WriteLine("++ STOP");
        
    	if (wThread != null)
    	{
    		done = true;
    	}
    }

    public void forward (int del, int psd)
    {
        Debug.WriteLine("Continuous walk - reverses if PSD < 25");

        fdel = del;
        fpsd = psd;
        
        if (wThread!= null)
        	stop();
        
	wThread = new Thread() {
	        public void run() 
	        {   	            
	            
	            int[][] step_r = reverse(step);
	            
	            int count = 0;
	            
	            int[][] nxt = step;
	            done = false;
	            first= true;
	            	            
                    while (done==false || count !=0)
                    {
                            if (count==0)
                            {
                                int d = readdistance();
                                Debug.WriteLine("PSD=" + d);

                                if (d > fpsd)
                                {
                                    nxt = step;
                                }
                                else
                                {
                                    nxt = step_r;
                                }
                            }
                    w.PlayPose(fdel, 1, cv18(nxt[count]), first);
                    if (first) first=false;
                    count++;
                    if (count>15) count=0;
                    }
	        }};
        wThread.start();
    }

    private int readdistance()
    {
        if (w.wckReadPos(30, 5))
        {
            return w.respnse[0];
        }
        return 0;
    }

}

